d2l-search-facets-grouping.html

```diff
@@ -0,0 +1,135 @@
+			:host {
+				display: flex;
+				flex-direction: column;
+			}
+			.d2l-search-facets-grouping-label {
+				@apply --d2l-body-compact-text;
+				font-weight: bold;
+			}
+
+			.d2l-search-facets-grouping-show-more > button {
+				@apply --d2l-body-compact-text;
+				border: none;
+				color: var(--d2l-color-celestine);
+				font-family: inherit;
+				outline: none;
+				padding: 0px;
+			}
+
+			.d2l-search-facets-grouping-show-more > button:hover,
+			.d2l-search-facets-grouping-show-more > button:focus {
+				cursor: pointer;
+				font-weight: bold;
+			}
+
+		</style>
+		<span class="d2l-search-facets-grouping-label">[[label]]</span>
+		<d2l-input-checkbox-spacer></d2l-input-checkbox-spacer>
+		<slot></slot>
+		<template is="dom-if" if="{{_showMoreButtonVisible}}">
+			<d2l-input-checkbox-spacer class="d2l-search-facets-grouping-show-more">
+				<button on-click="_showMore">More</button>
+			</d2l-input-checkbox-spacer>
+		</template>
+	</template>
+	<script>
+		/**
+		 * `<d2l-search-facets-grouping>`
+		 * A grouping of search facet options that keeps tracks of selected options
+		 */
+		class SearchFacetsGrouping extends Polymer.Element {
+			static get is() { return 'd2l-search-facets-grouping'; }
+			static get properties() {
+				return {
+					/**
+					* List of <d2l-search-facets-option> elements
+					*/
+					_optionElements: {
+						type: Array,
+						value: function() { return []; },
+					},
+					/**
+					* List of selected options (e.g., ['optionA', ...]s)
+					*/
+					_optionsSelected: {
+						type: Array,
+						value: function() { return []; },
+					},
+					/**
+					* Visibility of 'More' button - true if initiallyShown < _optionElements.length
+					*/
+					_showMoreButtonVisible: {
+						type: Boolean,
+						value: false,
+					},
+					/**
+					* Label of the grouping
+					*/
+					label: {
+						type: String,
+						value: '',
+					},
+					/**
+					* Number of facet options initially displayed. All options are shown by default
+					*/
+					initiallyShown: {
+						type: Number,
+					}
+				};
+			}
+
+			connectedCallback() {
+				super.connectedCallback();
+				this.addEventListener('d2l-search-facets-option-change', function(e) {
+					e.detail.checked
+						? this._optionsSelected.push(e.detail.option)
+						: this._optionsSelected.splice(this._optionsSelected.indexOf(e.detail.option), 1);
+
+					this.dispatchEvent(new CustomEvent(
+						'd2l-search-facets-grouping-change', {
+							bubbles: true, composed: true, detail:
+							{
+								grouping: this.label,
+								selected: this._optionsSelected,
+							}
+						}
+					));
+				}.bind(this));
+
+				// Child elements may not be rendered if this is a <dom-repeat>. Wait for next
+				// render to query for them
+				Polymer.RenderStatus.afterNextRender(this, function() {
+					this._optionElements = Polymer.FlattenedNodesObserver.getFlattenedNodes(this)
+						.filter(function(n) { return n.nodeType === Node.ELEMENT_NODE; });
+
```

Comment:
I wonder if you could use the `nth-child` CSS selector to do some really clever things here and avoid all this code. Specifically, I'm thinking of one of [the recipes](https://css-tricks.com/useful-nth-child-recipies/) that goes something like:

```css
// select all but the first 5
:nth-child(n+6) {
    display: none;
}
:host([all-visible]) :nth-child(n+6) {
    display: inline-block;
}
```

Browser support for that selector would be something to check on, and since the number itself is dynamic you may have to use JavaScript to add the CSS. But it could be cleaner to rely solely on CSS instead of needing to find all the children and update the visibility again later via JS.

Reply:
Yeah you'd need to dynamically add the rule via JavaScript (and remove it and re-add if the attribute value changes). I don't have a ton of experience doing this so you may need to experiment!

Reply:
That's what I figured you'd need. If you keep a reference to the `<style>` element, can you then just remove it when they click the "more" button"? You could also remove and re-add it if the `hide-after` attribute value changes.

Reply:
Also, since this is Polymer 2 and will get transpiled, you can use the backtick strings instead of needing to concatenate 💥

Reply:
Yeah, that's what I was thinking of doing 👍 
Cool, I'll definitely use template strings instead then

Reply:
Just wondering, did you have a use case in mind where the `hide-after` attribute would change after it's initially set?

Reply:
Personally - I remember when we were chatting about this at the whiteboard.

The 2 scenarios for "Hide after":
- (1) There's a TON of facet options (like 500 departments) and showing them all would be a ton of work
- (2) There's a few facet options, but only a couple have actual values

Ex2:
- [ ] English (10)
- [ ] French (2)
- MORE
- [ ] Arabic (0)
- [ ] German (0)
- Etc.

Seems like we're covering situation 1 very well with the hide after, but not situation 2. Based on the situations, I'm not sure if there's times where we would need to dynamically update the `hide-after`... unless we're implementing the behaviour of situation 2 through the `hide-after` property and back end stuff.

Reply:
I think adding a style block dynamically will cause a reflow similar to the original code.   Technically the new style block should be scoped as well, and I suspect shadyCSS would not scope this without some extra call to shadyCSS.  I'm not sure if this gets us anywhere.

Reply:
@cheics-d2l I think in situation 2 where the consumer could know those numbers up front, it can be set up (on the consumer side) so that those options wouldn't be in the list to be rendered

@dbatiste Not too familiar with shadyCSS but I'm assuming your concern is that adding it in this way to the shadowRoot isn't enough to keep it scoped to this element?

Reply:
Also... if there are like 500 departments that would show up, maybe we shouldn't be rendering them at all? That's a lot of HTML to send down and DOM to create if it's just going to be hidden initially. In this case, I'd want us to lazily load them when the user clicks "more"... but now that's going beyond what this web component can do, since it doesn't know anything about how to get those items.

Reply:
Regarding the handling of having many options - perhaps we could have an attribute (e.g., `remote-source`) that can be set on the grouping? (And `hide-after` wouldn't be set)

In this case, the code for adding/removing of CSS won't be run, but the 'More' button will be shown. Clicking the button will fire an event:
  - (`d2l-search-facet-grouping-more, { grouping: ...}`)

to tell the consumer to fetch more options

It will be up to the consumer to then add those elements (and also update the visibility of the 'More' button)

Reply:
I really hesitate to bake in too much logic like remote sources... at that point this simple component seems like it's doing too much.

What if we make it even simpler, where the group just has a single `bool` attribute called `has-more`. If that attribute is present, the more button is displayed and clicking it fires the event. It's then on the consumer to fetch more items and insert them as children of the group, and remove the `has-more` attribute if there aren't more.

Reply:
Yeah that's exactly what I meant 🙂 `remote-source` was the first thing that came to mind (as I used it elsewhere) - `has-more` sounds good to me

Reply:
Cool! Then really we can just remove all the auto hiding and `hide-after` stuff altogether I think, since the consumer can just put the first N or whatever in the group and set the `has-more` attribute and be done with it.